packages/belos/tpetra/src/solvers/Belos_Tpetra_GmresSstep.hpp

```diff
@@ -44,10 +44,141 @@
 #include "Belos_Tpetra_UpdateNewton.hpp"
 
 #include "KokkosBlas3_trsm.hpp"
+#include "KokkosBlas3_gemm.hpp"
+#include "KokkosBlas1_axpby.hpp"
 
 namespace BelosTpetra {
 namespace Impl {
 
+//
+// Compute upper-triangular matrix of SPD matrix, based on Cholesky or SVD
+template<class SC = Tpetra::Operator<>::scalar_type>
+class Chol {
+private:
+  bool useSVQR;
+
+public:
+  using LO = int;
+  using STS = Teuchos::ScalarTraits<SC>;
+  using mag_type = typename STS::magnitudeType;
+  using STM = Teuchos::ScalarTraits<mag_type>;
+  using dense_matrix_type = Teuchos::SerialDenseMatrix<LO, SC>;
+  using dense_vector_type = Teuchos::SerialDenseVector<LO, SC>;
+  using real_vector_type = Teuchos::SerialDenseVector<LO, mag_type>;
+  using blas_type = Teuchos::BLAS<LO, SC>;
+  using lapack_type = Teuchos::LAPACK<LO, SC>;
+
+  /// \brief Default constructor
+  ///
+  Chol () :
+  useSVQR (false)
+  {}
+
+  Chol (bool useSVQR_) :
+  useSVQR (useSVQR_)
+  {}
+
+  int
+  factor (Teuchos::FancyOStream* outPtr, int ncols, SC* Rdata, int ldr)
+  {
+    blas_type blas;
+    lapack_type lapack;
+    const SC zero = STS::zero ();
+
+    int info = 0;
+    if (useSVQR) {
+      real_vector_type S (ncols);
+      dense_matrix_type VT (ncols, ncols);
+      mag_type *Sdata = S.values();
+      SC *VTdata = VT.values();
+
+      // figure out workspace size
+      SC work;
+      mag_type rwork;
+      int ione = 1;
+      int ldvt = int (VT.stride());
+      lapack.GESVD('N', 'A', ncols, ncols, Rdata, ldr, Sdata, &work, 1, 
+                   VTdata, ldvt, &work, -ione, &rwork, &info);
+      int lwork = Teuchos::as<int> (STS::real (work));
+      lapack.GEQRF(ncols, ncols, Rdata, ldr, &work, &work, -ione, &info);
+      lwork = (lwork > Teuchos::as<int> (STS::real (work)) ? lwork : Teuchos::as<int> (STS::real (work)));
+
+      // diagonal scaling
+      dense_vector_type D (ncols);
+      bool diag_scale = true;
+      if (diag_scale) {
+        for (int i=0; i<ncols; i++) {
+          D(i) = std::sqrt(Rdata[i + i*ldr]);
+          for (int j=0; j<ncols; j++) {
+            Rdata[i + j*ldr] /= D(i);
+            Rdata[j + i*ldr] /= D(i);
+          }
+        }
+      }
+
+      // compute SVD
+      dense_vector_type W (lwork);
+      SC *Wdata = W.values();
+      lapack.GESVD('N', 'A', ncols, ncols, Rdata, ldr, Sdata, &work, 1,
+                   VTdata, ldvt, Wdata, lwork, &rwork, &info);
+      TEUCHOS_TEST_FOR_EXCEPTION
+        (info != 0, std::runtime_error, " GmresSstep::GESVD returned info = " << info << " with ncols = " << ncols);
+
+      // Repllace small singular values to tol
+      bool replace_small_s = true;
```

Comment:
I am certain we should be replacing the small singular values. I just left it as a boolean option since I wanted to double-check with the new features.